File: app/schema/dict.py
"""数据字典 Schema：字典类型与字典项的创建/更新/输出/分页。

字典为全局配置（不绑定项目/部门），时间字段统一以「北京时间墙钟」字符串对外。
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_serializer

from app.core.clock import LOCAL_TZ


class DictItemBase(BaseModel):
    label: str = Field(..., description="显示名称")
    value: str = Field(..., description="存储值")
    sort: int = Field(0, description="排序(升序)")
    enabled: bool = True
    remark: str | None = None
    ext: str | None = Field(None, description="扩展字段(颜色/图标等)")


class DictItemCreate(DictItemBase):
    pass


class DictItemUpdate(BaseModel):
    label: str | None = None
    value: str | None = None
    sort: int | None = None
    enabled: bool | None = None
    remark: str | None = None
    ext: str | None = None


class DictItemOut(DictItemBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    type_code: str
    created_at: datetime | None = None
    updated_at: datetime | None = None

    @field_serializer("created_at", "updated_at")
    def _serialize_dt(self, v: datetime | None) -> str | None:
        if v is None:
            return None
        return v.astimezone(LOCAL_TZ).replace(tzinfo=None).isoformat()


class DictTypeCreate(BaseModel):
    code: str = Field(..., description="类型编码(唯一)")
    name: str = Field(..., description="类型名称")
    description: str | None = None
    items: list[DictItemCreate] = Field(default_factory=list, description="初始字典项")


class DictTypeUpdate(BaseModel):
    name: str | None = None
    description: str | None = None


class DictTypeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    name: str
    description: str | None = None
    system: bool = False
    items: list[DictItemOut] = Field(default_factory=list)
    created_at: datetime | None = None
    updated_at: datetime | None = None

    @field_serializer("created_at", "updated_at")
    def _serialize_dt(self, v: datetime | None) -> str | None:
        if v is None:
            return None
        return v.astimezone(LOCAL_TZ).replace(tzinfo=None).isoformat()


class DictTypePage(BaseModel):
    total: int
    items: list[DictTypeOut]
    page: int = 1
    size: int = 20

File: web/src/api/dict.ts
// 数据字典 API 封装
import { http } from "@/utils/request";

export interface DictItem {
  id: number;
  type_code: string;
  label: string;
  value: string;
  sort: number;
  enabled: boolean;
  remark: string | null;
  ext: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface DictType {
  id: number;
  code: string;
  name: string;
  description: string | null;
  system: boolean;
  items: DictItem[];
  created_at: string | null;
  updated_at: string | null;
}

export interface DictTypePage {
  total: number;
  items: DictType[];
  page: number;
  size: number;
}

export interface DictTypeCreate {
  code: string;
  name: string;
  description?: string | null;
  items?: DictItemCreate[];
}

export interface DictItemCreate {
  label: string;
  value: string;
  sort?: number;
  enabled?: boolean;
  remark?: string | null;
  ext?: string | null;
}

export type DictItemUpdate = Partial<DictItemCreate>;

// 字典类型列表
export function fetchDictTypes(params: {
  keyword?: string;
  page?: number;
  size?: number;
}): Promise<DictTypePage> {
  return http<DictTypePage>({ url: "/v1/dicts", method: "get", params });
}

// 类型详情
export function fetchDictType(code: string): Promise<DictType> {
  return http<DictType>({ url: `/v1/dicts/${code}`, method: "get" });
}

// 创建字典类型
export function createDictType(payload: DictTypeCreate): Promise<DictType> {
  return http<DictType>({ url: "/v1/dicts", method: "post", data: payload });
}

// 更新字典类型
export function updateDictType(
  code: string,
  payload: { name?: string; description?: string | null },
): Promise<DictType> {
  return http<DictType>({ url: `/v1/dicts/${code}`, method: "put", data: payload });
}

// 删除字典类型
export async function deleteDictType(code: string): Promise<void> {
  await http({ url: `/v1/dicts/${code}`, method: "delete" });
}

// 字典项列表（enabledOnly=true 供业务下拉引用）
export function fetchDictItems(code: string, enabledOnly = false): Promise<DictItem[]> {
  return http<DictItem[]>({
    url: `/v1/dicts/${code}/items`,
    method: "get",
    params: { enabled_only: enabledOnly },
  });
}

// 新增字典项
export function createDictItem(code: string, payload: DictItemCreate): Promise<DictItem> {
  return http<DictItem>({ url: `/v1/dicts/${code}/items`, method: "post", data: payload });
}

// 更新字典项
export function updateDictItem(itemId: number, payload: DictItemUpdate): Promise<DictItem> {
  return http<DictItem>({ url: `/v1/dicts/items/${itemId}`, method: "put", data: payload });
}

// 删除字典项
export async function deleteDictItem(itemId: number): Promise<void> {
  await http({ url: `/v1/dicts/items/${itemId}`, method: "delete" });
}
